[visionOS] Refine activation rules for Image Fullscreen

etiennesegonzac opened the pull request:
#### 4ac0a7c004e0484efe684fc41487727bb9b0c5c7
<pre>
[visionOS] Refine activation rules for Image Fullscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=272218">https://bugs.webkit.org/show_bug.cgi?id=272218</a>
&lt;<a href="https://rdar.apple.com/125913434">rdar://125913434</a>&gt;

Reviewed by Said Abou-Hallawa.

Introduce a `shouldUseQuickLookForFullscreen()` method on Images. It
defaults to false and queries the ImageDecoder for bitmap images.

Gate the use of Image Fullscreen on this flag.

* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::shouldUseQuickLookForFullscreen const):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::shouldUseQuickLookForFullscreen const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::shouldUseQuickLookForFullscreen const):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::shouldUseQuickLookForFullscreen const):
Wire the `shouldUseQuickLookForFullscreen()` check to the ImageDecoder.

* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::shouldUseQuickLookForFullscreen const):
Add the WKA entry point.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
Add the `shouldUseQuickLookForFullscreen()` check.

Canonical link: <a href="https://commits.webkit.org/277177@main">https://commits.webkit.org/277177@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8edeb5cdb116fb7fdfc3d8b32fd73d23c166bf4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46788 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25949 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49411 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49465 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42835 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49095 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30347 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23415 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38121 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22954 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40298 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19417 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20284 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41437 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4834 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43033 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41803 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51338 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21797 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18151 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45406 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23085 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44365 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10359 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23583 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22794 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/BitmapImageSource.h:
```diff
@@ -70,6 +70,10 @@ class BitmapImageSource : public ImageSource {
 
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+    bool shouldUseQuickLookForFullscreen() const;
+#endif
```
I think this cab be private.

shallawa commented on Source/WebCore/platform/graphics/BitmapImageSource.cpp:
```diff
@@ -743,6 +743,13 @@ RepetitionCount BitmapImageSource::repetitionCount() const
+{
+    return m_decoder->shouldUseQuickLookForFullscreen();
+}
+#endif
```
Do we need to cache the return of this metadata? If yes, you can follow the pattern of the `orientation()` if you are getting it from the first frame metadata. Or you can follow the pattern of `uti()` if you are getting it from the image metadata.

shallawa commented on Source/WebCore/platform/graphics/BitmapImage.h:
```diff
@@ -76,6 +76,9 @@ class BitmapImage final : public Image {
 #endif
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+    bool shouldUseQuickLookForFullscreen() const final { return m_source->shouldUseQuickLookForFullscreen(); }
+#endif
```
I think this can be private (at least for this patch) since you are using it through the `Image` not the `BitmapImage`.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
```diff
@@ -75,6 +75,10 @@ class ImageDecoderCG final : public ImageDecoder {
 
     static String decodeUTI(CGImageSourceRef, const SharedBuffer&);
 
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+    bool shouldUseQuickLookForFullscreen() const;
+#endif
```
I think this can be private since you are calling it through `ImageDecoder` not `ImageDecoderCG`.

shallawa commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -230,7 +230,7 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element,
 
             auto* image = renderImage->cachedImage()->image();
-            if (!image || !image->isBitmapImage())
+            if (!image || !image->isBitmapImage() || !image->shouldUseQuickLookForFullscreen())
```
I think we can remove `!image->isBitmapImage()` now since `shouldUseQuickLookForFullscreen()` will return false for non `BitmapImage` images.

shallawa commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -246,11 +246,12 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element,
-                mimeType = MIMETypeRegistry::mimeTypeForPath(cachedImage->url().string());
+                if (!MIMETypeRegistry::isSupportedImageMIMEType(mimeType))
+                    mimeType = MIMETypeRegistry::mimeTypeForExtension(image->filenameExtension());
+            }
```
I think this part is not related to this PR. Please remove and address it in separate PR.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -45,6 +45,19 @@
+#endif // QUICKLOOK_FULLSCREEN
+
+}
+#endif
```
I think it is better to move this part to the end of this source file.

etiennesegonzac committed:
[visionOS] Refine activation rules for Image Fullscreen
https://bugs.webkit.org/show_bug.cgi?id=272218
<rdar://125913434>

Reviewed by Said Abou-Hallawa.

Introduce a `shouldUseQuickLookForFullscreen()` method on Images. It
defaults to false and queries the ImageDecoder for bitmap images.

Gate the use of Image Fullscreen on this flag.

* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::shouldUseQuickLookForFullscreen const):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::shouldUseQuickLookForFullscreen const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::shouldUseQuickLookForFullscreen const):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::shouldUseQuickLookForFullscreen const):
Wire the `shouldUseQuickLookForFullscreen()` check to the ImageDecoder.

* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::shouldUseQuickLookForFullscreen const):
Add the WKA entry point.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
Add the `shouldUseQuickLookForFullscreen()` check.

Canonical link: https://commits.webkit.org/277177@main
